utils/prometheus/src/lib.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use futures_util::{FutureExt, future::Future};
+use hyper::http::StatusCode;
+use hyper::{Server, Body, Request, Response, service::{service_fn, make_service_fn}};
+use prometheus::{Encoder, TextEncoder, core::Collector};
+use std::net::SocketAddr;
+#[cfg(not(target_os = "unknown"))]
+mod networking;
+
+pub use prometheus::{
+	Registry, Error as PrometheusError, Opts,
+	core::{
+		GenericGauge as Gauge, GenericCounter as Counter,
+		GenericGaugeVec as GaugeVec, GenericCounterVec as CounterVec,
+		AtomicF64 as F64, AtomicI64 as I64, AtomicU64 as U64,
+	}
+};
+
+pub fn register<T: Clone + Collector + 'static>(metric: T, registry: &Registry) -> Result<T, PrometheusError> {
+	registry.register(Box::new(metric.clone()))?;
+	Ok(metric)
+}
+
+#[derive(Debug, derive_more::Display, derive_more::From)]
+pub enum Error {
+	/// Hyper internal error.
+	Hyper(hyper::Error),
+	/// Http request error.
+	Http(hyper::http::Error),
+	/// i/o error.
+	Io(std::io::Error),
+	#[display(fmt = "Prometheus exporter port {} already in use.", _0)]
+	PortInUse(SocketAddr)
+}
+
+impl std::error::Error for Error {
+	fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+		match self {
+			Error::Hyper(error) => Some(error),
+			Error::Http(error) => Some(error),
+			Error::Io(error) => Some(error),
+			Error::PortInUse(_) => None
+		}
+	}
+}
+
+async fn request_metrics(req: Request<Body>, registry: Registry) -> Result<Response<Body>, Error> {
+	if req.uri().path() == "/metrics" {
+		let metric_families = registry.gather();
+		let mut buffer = vec![];
+		let encoder = TextEncoder::new();
+		encoder.encode(&metric_families, &mut buffer).unwrap();
+
+		Response::builder()
+			.status(StatusCode::OK)
+			.header("Content-Type", encoder.format_type())
+			.body(Body::from(buffer))
+			.map_err(Error::Http)
+	} else {
+		Response::builder()
+		.status(StatusCode::NOT_FOUND)
```

Comment:
This line and the ones below would need to be indented. See `Response::builder().status(...)` above.

Reply:
complete